doc/reference/trd-syscalls.md

```diff
@@ -0,0 +1,715 @@
+
+The Exit system call class is how a userspace process terminates. 
+Successful calls to Exit system calls do not return.
+
+There are two Exit system calls:
+  - `exit-terminate`
+  - `exit-restart`
+
+The first call, `exit-terminate`, terminates the process and tells the
+kernel that it may reclaim and reallocate the process as well as all of its
+resources. Usually this indicates that the process has completed
+its work. 
+
+The second call, `exit-restart`, terminates the process and tells the kernel
+that the application would like to restart if possible. If the kernel 
+restarts the application, it MUST assign it a new process identifier. The
+kernel MAY reuse existing process resources (e.g., RAM regions) or MAY
+allocate new ones.
+
+The register arguments for Exit system calls are as follows. The registers
+r0-r3 correspond to r0-r3 on CortexM and a0-a3 on RISC-V.
+
+| Argument          | Register |
+|-------------------|----------|
+| Exit identifer    | r0       |
+| Completion code   | r1       |
+
+The exit identifier specifies which call is invoked.
+
+| System call     | Exit identifier value |
+|-----------------|-----------------------|
+| exit-terminate  |                     0 |
+| exit-restart    |                     1 |
+
+The difference between `exit-terminate` and `exit-restart` is what behavior
+the application asks from the kernel. With `exit-terminate`, the application
+tells the kernel that it considers itself completed and does not need to run
+again. With `exit-restart`, it tells the kernel that it would like to be
+rebooted and run again. For example, `exit-terminate` might be used by a 
+process that stores some one-time data on flash, while `exit-restart` might
+be used if the process runs out of memory.
+
+The completion code is an unsigned 32-bit number which indicates status. This
+information can be stored in the kernel and used in management or policy decisions.
+The definition of these status codes is outside the scope of this document.
+
+If an exit syscall is successful, it does not return. Therefore, the return
+value of an exit syscall is always _Failure_. `exit-restart` and 
+`exit-terminate` MUST always succeed and so never return. 
+
+5 Userspace Library Methods
+=================================
+
+This section describes the method signatures for system calls and callbacks in C and Rust.
+Because C allows a single return value but Tock system calls can return multiple values,
+the low-level APIs are not idiomatic C. These low-level APIs are translated into standard C
+code by the userspace library.
+
+5.1 libtock-c
+---------------------------------
+
+5.1 libtock-rs
+---------------------------------
+
+6 The Driver Trait
+=================================
+
+The core kernel, in response to userspace system calls, invokes methods on the Driver
+trait implemented by system call drivers.  This section describes the Driver trait API
+and how it interacts with the core kernel's system calls. Note that
```

Comment:
TODO? Do we want to flesh this out or just drop it from this doc?

Reply:
#2431